java-spiffe-core/src/main/java/io/spiffe/spiffeid/SpiffeId.java

```diff
@@ -0,0 +1,94 @@
+package io.spiffe.spiffeid;
+
+import lombok.NonNull;
+import lombok.Value;
+import lombok.val;
+import org.apache.commons.lang3.StringUtils;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a SPIFFE ID as defined in SPIFFE standard.
+ * <p>
+ * @see <a href="https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md">https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md</a>
+ */
+@Value
+public class SpiffeId {
+
+    public static final String SPIFFE_SCHEME = "spiffe";
+
+    TrustDomain trustDomain;
+
+    String path;
+
+    private SpiffeId(final TrustDomain trustDomain, final String path) {
+        this.trustDomain = trustDomain;
+        this.path = path;
+    }
+
+    /**
+     * Returns an instance representing a SPIFFE ID, containing the trust domain and
+     * a path generated joining the segments (e.g. /path1/path2).
+     *
+     * @param trustDomain an instance of a {@link TrustDomain}
+     * @param segments    a list of string path segments
+     *
+     * @return a {@link SpiffeId}
+     */
+    public static SpiffeId of(@NonNull final TrustDomain trustDomain, final String... segments) {
+        val path = Arrays.stream(segments)
+                .filter(StringUtils::isNotBlank)
+                .map(SpiffeId::normalize)
+                .map(s -> '/' + s)
+                .collect(Collectors.joining());
+        return new SpiffeId(trustDomain, path);
+    }
+
+    /**
+     * Parses a SPIFFE ID from a string (e.g. spiffe://example.org/test).
+     *
+     * @param spiffeIdAsString a String representing a SPIFFE ID
+     * @return A {@link SpiffeId}
+     * @throws IllegalArgumentException if the given string cannot be parsed
+     */
+    public static SpiffeId parse(@NonNull final String spiffeIdAsString) {
```

Comment:
This validation accepts some strings that are not considered valid SPIFFE IDs by the specification. The following strings are not valid SPIFFE IDs, but would currently be considered valid by this method:

```
spiffe://foo.com/bar#invalidfragment
spiffe://foo.com/bar?invalidQuery
spiffe://user@foo.com/bar
```
SPIFFE IDs are also required to be at most 2048 bytes in length, and the trust domain must be at most 255 bytes in length.

Reply:
definitely many of the validations are missing. Now I added the validation required by the standard and the tests.  Thanks for catching this.

Reply:
The spec guidance around not generating SPIFFE IDs longer than 2048  and HOST longer than 255 is a SHOULD NOT, meaning that it allows a SPIFFE implementation to deviate at the risk of interoperability.  Our library should still work with those implementations since even if they deviate, they are still within the spec.